### PR TITLE
PS-7382: Rewrite Hashicorp Vault implementation and implement non-root tokens

### PR DESCRIPTION
**Pull request by Sudokamikaze**
Builds: https://ps57.cd.percona.com/job/percona-server-5.7-param-ps-7382/6/

### CODE REVIEW

**Review by EvgeniyPatlan — APPROVED** [review]

File: docker/run-test [hunk 1]
```diff
@@ -11,11 +11,39 @@ if [[ ${CI_FS_MTR} == 'yes' ]]; then
     CI_FS_DOCKER_FLAG="--mount type=bind,source=/mnt/ci_disk_dir_${CMAKE_BUILD_TYPE},destination=/tmp/ci_disk_dir"
 fi
 
+if [[ "${KEYRING_VAULT_MTR}" == 'yes' ]] && [[ ${DOCKER_OS} != 'centos:6' ]] && [[ ${DOCKER_OS} != 'i386/centos:6' ]]; then
+    if [[ $(docker network list | grep bridge-vault) != *bridge-vault* ]]; then
+        docker network create bridge-vault
+    fi
+
+    # Bootstrap dev mode vault instances
+    BOOTSTRAP_DEV_V1="$(bash $SCRIPTS_DIR/bootstrap-vault --mode=dev --version="0.9.6" --alias=v1 --port="9200" --devtoken=${VAULT_V1_DEV_TOKEN})"
```
Comment (percona-ysorokin) [inline docker/run-test]: Please, extract `0.9.6` and `1.5.4` into bash script variables as it used to be in the previous implementation.
Also change `1.5.4` to `1.6.2` as it is the newest version available.

File: local/bootstrap-vault [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+api_addr = "https://vault-prod-${ALIAS}:${PORT}"
+cluster_addr = "https://vault-prod-${ALIAS}:${cluster_addr_port}"
+EOF
+
+cat <<-EOF | tee /tmp/vault.d-${ALIAS}-prod/policies/mtr.hcl
+path "mtr/*" {
+  capabilities = ["create", "read", "update", "delete", "list"]
+}
+
+path "sys/mounts" {
```
Comment (percona-ysorokin) [inline local/bootstrap-vault]: Please, remove `sys/mounts` and `sys/mounts/*` sections

File: local/bootstrap-vault [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+function prepare_dev_env {
+    mkdir -p /tmp/vault.d-${ALIAS}-dev/ /tmp/vault.d-${ALIAS}-dev/policies
+cat <<-EOF | tee /tmp/vault.d-${ALIAS}-dev/policies/mtr.hcl
+path "mtr/*" {
+  capabilities = ["create", "read", "update", "delete", "list"]
+}
+
+path "sys/mounts" {
```
Comment (percona-ysorokin) [inline local/bootstrap-vault]: Likewise

File: local/bootstrap-vault [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        VAULT_PROD_UNSEAL_KEYS=$(echo "$VAULT_PROD_KEYS" | jq '.unseal_keys_b64' | awk -F '"' '{print $2}' | xargs)
+        VAULT_PROD_TOKEN=$(echo "$VAULT_PROD_KEYS" | jq -r '.root_token')
+
+        for key in $VAULT_PROD_UNSEAL_KEYS; do
+            docker exec -e VAULT_ADDR="https://vault-prod-${ALIAS}:${PORT}" -e VAULT_CACERT="/etc/vault.d/ssl/ca.pem" -t vault-prod-${ALIAS} vault operator unseal "$key"
+            sleep 1
+        done
+
+        docker exec -e VAULT_TOKEN=${VAULT_PROD_TOKEN} -e VAULT_ADDR="https://vault-prod-${ALIAS}:${PORT}" -e VAULT_CACERT="/etc/vault.d/ssl/ca.pem" -t vault-prod-${ALIAS} vault policy write mtr /etc/vault.d/policies/mtr.hcl
+        MTR_TOKEN_TMP=$(docker exec -e VAULT_TOKEN=${VAULT_PROD_TOKEN} -e VAULT_ADDR="https://vault-prod-${ALIAS}:${PORT}" -e VAULT_CACERT="/etc/vault.d/ssl/ca.pem" -t vault-prod-${ALIAS} vault token create -no-default-policy -format=json -policy=mtr)
```
Comment (percona-ysorokin) [inline local/bootstrap-vault:118]: I managed to simplify this by using single command
```
VAULT_MTR_TOKEN=$(vault token create -policy=mtr -field=token)
```

File: local/bootstrap-vault [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        VAULT_MTR_TOKEN=$(echo $MTR_TOKEN_TMP | jq -r '.auth.client_token')
+        echo "Production token: ${VAULT_PROD_TOKEN}"
+        echo "MTR token: ${VAULT_MTR_TOKEN}"
+    else
+        prepare_dev_env
+        docker run --cap-add=IPC_LOCK -d --name=vault-dev-${ALIAS} --hostname=vault-dev-${ALIAS} --net bridge-vault -v /tmp/vault.d-${ALIAS}-dev:/etc/vault.d vault:${VERSION} server -dev -dev-listen-address=vault-dev-${ALIAS}:${PORT} -dev-root-token-id=${DEVTOKEN}
+        sleep 3
+        docker exec -e VAULT_TOKEN=${DEVTOKEN} -e VAULT_ADDR="http://vault-dev-${ALIAS}:${PORT}" -t vault-dev-${ALIAS} vault policy write mtr /etc/vault.d/policies/mtr.hcl
+        MTR_TOKEN_TMP=$(docker exec -e VAULT_TOKEN=${DEVTOKEN} -e VAULT_ADDR="http://vault-dev-${ALIAS}:${PORT}" -t vault-dev-${ALIAS} vault token create -no-default-policy -format=json -policy=mtr)
+        VAULT_MTR_TOKEN=$(echo $MTR_TOKEN_TMP | jq -r '.auth.client_token')
```
Comment (percona-ysorokin) [inline local/bootstrap-vault:128]: Likewise

File: local/bootstrap-vault [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+MODE=dev
+VERSION=0.9.6
+DEVTOKEN="devtoken"
+ALIAS=v1
+PORT="9200"
+parse_arguments PICK-ARGS-FROM-ARGV "$@"
+
+main
```
Comment (percona-ysorokin) [inline local/bootstrap-vault]: no new line at the end of file

File: local/test-binary [hunk 7]
```diff
@@ -169,109 +169,69 @@ if [[ "${CI_FS_MTR}" = 'yes' ]]; then
...
-    MTR_VAULT_TOKEN=${VAULT_V1_DEV_TOKEN} \
+    MTR_VAULT_ADMIN_TOKEN=${VAULT_V1_DEV_TOKEN} \
+    MTR_VAULT_PLUGIN_TOKEN=${VAULT_V1_DEV_MTR_TOKEN} \
     MTR_VAULT_VERSION=1 \
     MTR_BUILD_THREAD=auto ./mtr \
         --force \
         --max-test-fail=0 \
         --parallel=$(nproc) \
         --retry=0 \
+        --big-test \
```
Comment (percona-ysorokin) [inline local/test-binary:189]: Please, also add `--unit-tests` to each `./mtr` execution

File: local/test-binary [hunk 8]
```diff
@@ -169,109 +169,69 @@ if [[ "${CI_FS_MTR}" = 'yes' ]]; then
...
-    MTR_VAULT_TOKEN=${VAULT_V2_DEV_TOKEN} \
+    MTR_VAULT_ADMIN_TOKEN=${VAULT_V2_DEV_TOKEN} \
+    MTR_VAULT_PLUGIN_TOKEN=${VAULT_V2_DEV_MTR_TOKEN} \
     MTR_VAULT_VERSION=2 \
     MTR_BUILD_THREAD=auto ./mtr \
         --force \
         --max-test-fail=0 \
         --parallel=$(nproc) \
         --retry=0 \
+        --big-test \
```
Comment (percona-ysorokin) [inline local/test-binary:204]: Likewise

File: local/test-binary [hunk 9]
```diff
@@ -169,109 +169,69 @@ if [[ "${CI_FS_MTR}" = 'yes' ]]; then
...
+    MTR_VAULT_ADMIN_TOKEN=${VAULT_V1_PROD_ROOT_TOKEN} \
+    MTR_VAULT_PLUGIN_TOKEN=${VAULT_V1_PROD_MTR_TOKEN} \
+    MTR_VAULT_CA="${WORKDIR_ABS}/vault-prod-v1-ca.pem" \
     MTR_VAULT_VERSION=1 \
     MTR_BUILD_THREAD=auto ./mtr \
         --force \
         --max-test-fail=0 \
         --parallel=$(nproc) \
         --retry=0 \
+        --big-test \
```
Comment (percona-ysorokin) [inline local/test-binary:221]: Likewise

File: local/test-binary [hunk 10]
```diff
@@ -169,109 +169,69 @@ if [[ "${CI_FS_MTR}" = 'yes' ]]; then
...
+    MTR_VAULT_ADMIN_TOKEN=${VAULT_V2_PROD_ROOT_TOKEN} \
+    MTR_VAULT_PLUGIN_TOKEN=${VAULT_V2_PROD_MTR_TOKEN} \
+    MTR_VAULT_CA="${WORKDIR_ABS}/vault-prod-v2-ca.pem" \
     MTR_VAULT_VERSION=2 \
     MTR_BUILD_THREAD=auto ./mtr \
         --force \
         --max-test-fail=0 \
         --parallel=$(nproc) \
         --retry=0 \
+        --big-test \
```
Comment (percona-ysorokin) [inline local/test-binary:237]: Likewise

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: docker/run-test (hunk 1 above)
Comment (Sudokamikaze) [reply docker/run-test]: Fixed

File: local/bootstrap-vault (hunk 2 above)
Comment (Sudokamikaze) [reply local/bootstrap-vault]: Fixed

File: local/bootstrap-vault (hunk 3 above)
Comment (Sudokamikaze) [reply local/bootstrap-vault]: Fixed

File: local/bootstrap-vault (hunk 6 above)
Comment (Sudokamikaze) [reply local/bootstrap-vault]: Fixed

File: local/test-binary (hunk 7 above)
Comment (Sudokamikaze) [reply local/test-binary:189]: Added

File: local/test-binary (hunk 8 above)
Comment (Sudokamikaze) [reply local/test-binary:204]: Added

File: local/test-binary (hunk 9 above)
Comment (Sudokamikaze) [reply local/test-binary:221]: Added

File: local/test-binary (hunk 10 above)
Comment (Sudokamikaze) [reply local/test-binary:237]: Added

**Comment by Sudokamikaze** [comment]
@percona-ysorokin Here's the fresh build: https://ps57.cd.percona.com/job/percona-server-5.7-pipeline-ps-7382/164/

Please notice, that `-field=token` doesnt work due to extracting from docker: https://ps57.cd.percona.com/job/percona-server-5.7-pipeline-ps-7382/163/

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Pull request merged**